### PR TITLE
Sass 3537

### PR DESCRIPTION
**Pull request by satheeshgottemukkula**
### Description
Please include a brief summary of the change you've made and the issue that it fixes.

Add a link to the relevant story in Jira
SASS-3537 API#1785 (for tax year 23-24) and API#1391 Integration - Tax year specific API changes

##### Before Reviewing
- [ ]  Have you pulled the branch down?
- [ ]  Have you assigned yourself to the PR?
- [ ]  Have you moved the task to “in review” on JIRA?
- [ ]  Have you checked to ensure all dependencies are up to date?
- [ ]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [ ]  Have you run the tests?
- [ ]  Have you run the journey tests?
- [ ]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [ ]  Have you run the journey tests? (where applicable)
- [ ]  Have you addressed warnings where appropriate?
- [x]  Have you rebased against the current version of main?
- [x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [ ]  Have you checked the PR Builder passes?

### CODE REVIEW

File: app/connectors/httpParsers/APIParser.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  def handleAPIError[Response](response: HttpResponse, statusOverride: Option[Int] = None): Either[ErrorModel, Response] = {
+
+    val status = statusOverride.getOrElse(response.status)
+
+    try {
+      val json = response.json
+
+      lazy val desError = json.asOpt[ErrorBodyModel]
+      lazy val desErrors = json.asOpt[ErrorsBodyModel]
```
Comment (jhkldev) [inline app/connectors/httpParsers/APIParser.scala:46]: These vals should be renamed to `apiError` and `apiErrors`

File: app/connectors/httpParsers/SubmittedGiftAidIfHttpParser.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import models.giftAid.SubmittedGiftAidModel
+import models.{ErrorModel}
+import play.api.Logging
+import play.api.http.Status._
+import uk.gov.hmrc.http.{HttpReads, HttpResponse}
+import utils.PagerDutyHelper.PagerDutyKeys._
+import utils.PagerDutyHelper.pagerDutyLog
+
+object SubmittedGiftAidIfHttpParser extends APIParser with Logging {
```
Comment (jhkldev) [inline app/connectors/httpParsers/SubmittedGiftAidIfHttpParser.scala:27]: Needs to be renamed to `GetAnnualIncomeSourcePeriodHttpParser`

File: test/config/MockAppConfig.scala [hunk 3]
```diff
@@ -16,12 +16,20 @@
...
-  override val auditingEnabled: Boolean = false
-  override val graphiteHost: String = "localhost"
-  override val desBaseUrl: String = "des"
   override val environment: String = "dev"
   override val authorisationToken: String = "someToken"
+  override val authorisationTokenKey: String = "someToken"
+  override val ifBaseUrl: String = "/if"
+  override val ifEnvironment: String = "dev"
+
+  override def authorisationTokenFor(apiVersion: String): String = ???
```
Comment (jhkldev) [inline test/config/MockAppConfig.scala:34]: This should be changed to `"someToken"`. If this was called then it'd throw a `NotImplementedError` exception.

**Review by jhkldev — CHANGES_REQUESTED** [review]

File: app/connectors/httpParsers/GetAnnualIncomeSourcePeriodHttpParser.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import models.giftAid.SubmittedGiftAidModel
+import models.{ErrorModel}
+import play.api.Logging
+import play.api.http.Status._
+import uk.gov.hmrc.http.{HttpReads, HttpResponse}
+import utils.PagerDutyHelper.PagerDutyKeys._
+import utils.PagerDutyHelper.pagerDutyLog
+
+object GetAnnualIncomeSourcePeriodHttpParser extends APIParser with Logging {
+    type GetAnnualIncomeSourcePeriod = Either[ErrorModel, SubmittedGiftAidModel]
```
Comment (jhkldev) [inline app/connectors/httpParsers/GetAnnualIncomeSourcePeriodHttpParser.scala:28]: Can `SubmittedGiftAidModel` be renamed to `AnnualIncomeSourcePeriod`?

File: app/services/SubmittedGiftAidService.scala [hunk 5]
```diff
@@ -16,16 +16,22 @@
...
+class SubmittedGiftAidService @Inject()(submittedGiftAidConnector: SubmittedGiftAidConnector,
+                                        getAnnualIncomeSourcePeriodConnector: GetAnnualIncomeSourcePeriodConnector ) {
+  def getSubmittedGiftAid(nino: String, taxYear: Int)(implicit hc: HeaderCarrier): Future[SubmittedGiftAidResponse] = {
+    if (taxYear.equals(2024)) {
+      getAnnualIncomeSourcePeriodConnector.getAnnualIncomeSourcePeriod(nino, taxYear, Some(false))
+    }
+    else {
+      submittedGiftAidConnector.getSubmittedGiftAid(nino, taxYear)
+    }
+  }
```
Comment (jhkldev) [inline app/services/SubmittedGiftAidService.scala:36]: Can this service and the controller which calls it also be renamed to better reflect the API name?

File: app/utils/TaxYearUtils.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package utils
+
+import play.api.Logging
+
+object TaxYearUtils extends Logging {
+
+  def convertSpecificTaxYear(taxYear: Int): String = {
+  s"${taxYear - 1}-${taxYear.toString takeRight 2}"
```
Comment (jhkldev) [inline app/utils/TaxYearUtils.scala:24]: The braces are unnecessary here and the code should be indented.

File: conf/application.conf [hunk 7]
```diff
@@ -111,6 +111,18 @@ microservice {
       port = 9303
       #This is the port for the income-tax-submission-stub
     }
+
+
+    integration-framework {
+
```
Comment (jhkldev) [inline conf/application.conf:117]: The newline can be removed.

**Review by jhkldev — CHANGES_REQUESTED** [review]

File: app/connectors/httpParsers/GetAnnualIncomeSourcePeriodHttpParser.scala (hunk 4 above)
Comment (satheeshgottemukkula) [reply app/connectors/httpParsers/GetAnnualIncomeSourcePeriodHttpParser.scala:28]: This is the existing model, which we will be using for new API too

**Review by jhkldev — APPROVED** [review]

**Review by jhkldev — APPROVED** [review]

**Pull request merged**